pkg/streamingpromql/benchmarks/comparison_test.go

```diff
@@ -0,0 +1,306 @@
+// Provenance-includes-license: Apache-2.0
+// Provenance-includes-copyright: The Prometheus Authors
+
+package benchmarks
+
+import (
+	"context"
+	"math"
+	"os"
+	"slices"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/grafana/dskit/flagext"
+	"github.com/grafana/dskit/kv"
+	"github.com/grafana/dskit/kv/consul"
+	"github.com/grafana/dskit/ring"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/dskit/test"
+	"github.com/grafana/dskit/user"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/parser"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/mimir/pkg/distributor"
+	"github.com/grafana/mimir/pkg/ingester"
+	"github.com/grafana/mimir/pkg/ingester/client"
+	"github.com/grafana/mimir/pkg/querier"
+	"github.com/grafana/mimir/pkg/querier/stats"
+	"github.com/grafana/mimir/pkg/streamingpromql"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+// This is based on the benchmarks from https://github.com/prometheus/prometheus/blob/main/promql/bench_test.go.
+func BenchmarkQuery(b *testing.B) {
+	// Important: the setup below must remain in sync with the setup done in tools/benchmark-query-engine.
+	q := createBenchmarkQueryable(b, MetricSizes)
+	cases := TestCases(MetricSizes)
+
+	opts := streamingpromql.NewTestEngineOpts()
+	standardEngine := promql.NewEngine(opts)
+	streamingEngine, err := streamingpromql.NewEngine(opts)
+	require.NoError(b, err)
+
+	// Important: the names below must remain in sync with the names used in tools/benchmark-query-engine.
+	engines := map[string]promql.QueryEngine{
+		"standard":  standardEngine,
+		"streaming": streamingEngine,
+	}
+
+	ctx := user.InjectOrgID(context.Background(), UserID)
+
+	// Don't compare results when we're running under tools/benchmark-query-engine, as that will skew peak memory utilisation.
+	skipCompareResults := os.Getenv("STREAMING_PROMQL_ENGINE_BENCHMARK_SKIP_COMPARE_RESULTS") == "true"
+
+	for _, c := range cases {
+		start := time.Unix(int64((NumIntervals-c.Steps)*intervalSeconds), 0)
+		end := time.Unix(int64(NumIntervals*intervalSeconds), 0)
+
+		b.Run(c.Name(), func(b *testing.B) {
+			if !skipCompareResults {
+				// Check both engines produce the same result before running the benchmark.
+				standardResult, standardClose := c.Run(ctx, b, start, end, interval, standardEngine, q)
+				streamingResult, streamingClose := c.Run(ctx, b, start, end, interval, streamingEngine, q)
+
+				requireEqualResults(b, standardResult, streamingResult)
+
+				standardClose()
+				streamingClose()
+			}
+
+			for name, engine := range engines {
+				b.Run(name, func(b *testing.B) {
+					for i := 0; i < b.N; i++ {
+						res, cleanup := c.Run(ctx, b, start, end, interval, engine, q)
+
+						if res != nil {
+							cleanup()
+						}
+					}
+				})
+			}
+		})
+	}
+}
+
+func TestBenchmarkQueries(t *testing.T) {
```

Comment:
[nit] To me, from the name `TestBenchmarkQueries` looks a bit hard to understand what it does. To my understand this is a correctness that, to ensure standard engine returns result equal to the streaming one. I suggest to rename the test to clarify it (even at the cost of making it more verbose).

Reply:
Makes sense, I've renamed it to `TestBothEnginesReturnSameResultsForBenchmarkQueries`.